### PR TITLE
SSM & SecretsManager - GetParameters integration (with caveat)

### PR DESCRIPTION
**Pull request by jellybeat**
Patched SSM so **GetRequests** requests can support having SecretsManager secres in it. Also, slightly modified the method so when an inexistent secret is requested a 400 is managed instead of having the 500 thrown.

**NOTE:** I'm not much acquainted with Python so sorry if I'm not doing some things the most recommended Python way

## Caveat
The patch works as expected when making requests through the aws-cli or HTTP request (with postman) like so:

1) Create the param and secret

```
aws --endpoint-url http://localstack:4566 secretsmanager create-secret --name test_secret --secret-string 'test_secret' && \
aws --endpoint-url http://localstack:4566 ssm put-parameter --name test_parameter --value 'test_parameter'
```

2) Get existent and non existent params and secrets

```
aws --endpoint-url http://localstack:4566 ssm get-parameters \
    --names /aws/reference/secretsmanager/test_secret /aws/reference/secretsmanager/no_secret \
            test_parameter no_parameter
```

Response:

```
{
    "Parameters": [
        {
            "Name": "/aws/reference/secretsmanager/test_secret",
            "Type": "SecureString",
            "Value": "test_secret",
            "SourceResult": {
                "ARN": "arn:aws:secretsmanager:us-east-1:1234567890:secret:test_secret-HHFCE",
                "Name": "test_secret",
                "VersionId": "d9dd4abe-66aa-491e-b124-63258e7c33cc",
                "SecretString": "test_secret",
                "VersionStages": [
                    "AWSCURRENT"
                ],
                "CreatedDate": "2020-11-15T20:54:28.000Z"
            },
            "LastModifiedDate": "2020-11-15T20:54:28+00:00"
        },
        {
            "Name": "test_parameter",
            "Value": "test_parameter",
            "Version": 1,
            "LastModifiedDate": "2020-11-15T20:54:29.656000+00:00",
            "ARN": "arn:aws:ssm:us-east-1:1234567890:parameter/test_parameter"
        }
    ],
    "InvalidParameters": [
        "/aws/reference/secretsmanager/no_secret",
        "no_parameter"
    ]
}
```

**BUT**
When requesting through an SDK (Java in this case) there are problems in parsing the JSON response. I would like to request some help as of why this is the case. I believe that is because when secrets are involved, the response is sent directly in the listener instead of returning a Request.

```
Caused by: java.lang.RuntimeException: We expected a VALUE token but got: START_OBJECT
  at com.amazonaws.transform.JsonUnmarshallerContextImpl.readCurrentJsonTokenValue(JsonUnmarshallerContextImpl.java:152)
  at com.amazonaws.transform.JsonUnmarshallerContextImpl.readText(JsonUnmarshallerContextImpl.java:135)
  at com.amazonaws.transform.SimpleTypeJsonUnmarshallers$StringJsonUnmarshaller.unmarshall(SimpleTypeJsonUnmarshallers.java:34)
  at com.amazonaws.transform.SimpleTypeJsonUnmarshallers$StringJsonUnmarshaller.unmarshall(SimpleTypeJsonUnmarshallers.java:32)
... and so on
```

┆Issue is synchronized with this [Jira Bug](https://localstack.atlassian.net/browse/LOC-323) by [Unito](https://www.unito.io/learn-more)

### CODE REVIEW

File: localstack/services/ssm/ssm_listener.py [hunk 1]
```diff
@@ -19,12 +19,44 @@ def normalize_name(param_name):
...
+        del secret_info['ResponseMetadata']
+        result = {'Parameter': {'SourceResult': secret_info, 'Name': name, 'Value':
+                secret_info.get('SecretString'), 'Type': 'SecureString',
                             'LastModifiedDate': secret_info.get('CreatedDate')}}
+        return result
+    except client.exceptions.ResourceNotFoundException:
+        return None
+
+
+def has_secrets(names):
```
Comment (whummer) [inline localstack/services/ssm/ssm_listener.py:33]: nitpick: perhaps a bit more "pythonic":
```
def has_secrets(names):
    for name in names:
        if name.startswith('/aws/reference/secretsmanager'):
            return True
```
(not need to return False - by default `None` is returned which is a falsy value in Python.)

File: localstack/services/ssm/ssm_listener.py [hunk 2]
```diff
@@ -19,12 +19,44 @@ def normalize_name(param_name):
...
+            break
+    return hasSecrets
+
+
+def get_params_and_secrets(names):
+    ssmClient = aws_stack.connect_to_service('ssm')
+    result = {'Parameters': [], 'InvalidParameters': []}
+
+    for name in names:
+        if name[0:29] == '/aws/reference/secretsmanager':
```
Comment (whummer) [inline localstack/services/ssm/ssm_listener.py:47]: better use
```
if name.startswith('/aws/reference/secretsmanager'):
    ...
```

File: localstack/services/ssm/ssm_listener.py [hunk 3]
```diff
@@ -19,12 +19,44 @@ def normalize_name(param_name):
...
+    hasSecrets = False
+    for name in names:
+        if name[0:29] == '/aws/reference/secretsmanager':
+            hasSecrets = True
+            break
+    return hasSecrets
+
+
+def get_params_and_secrets(names):
+    ssmClient = aws_stack.connect_to_service('ssm')
```
Comment (whummer) [inline localstack/services/ssm/ssm_listener.py:43]: nitpick: let's better use `snake_case` for variable names, as in the rest of the code base (i.e., `ssm_client`).

File: tests/integration/test_ssm.py [hunk 4]
```diff
@@ -53,6 +53,42 @@ def test_get_paramter(self):
...
+        )
+
+        complete_secret = secret_path + secret_name
+        response = ssm_client.get_parameters(Names=[param_name, complete_secret,
+            'inexistent_param', secret_path + 'inexistent_secret'])
+        found = response.get('Parameters')
+        notFound = response.get('InvalidParameters')
+
+        for param in found:
+            self.assertTrue(param['Name'] in [param_name, complete_secret])
```
Comment (whummer) [inline tests/integration/test_ssm.py:88]: nit: better use:
```
self.assertIn(param['Name'], [param_name, complete_secret])
```
(same for line 90 below)

File: tests/integration/test_ssm.py [hunk 5]
```diff
@@ -53,6 +53,42 @@ def test_get_paramter(self):
...
+            Name=secret_name,
+            SecretString='my_secret',
+            Description='testing creation of secrets'
+        )
+
+        complete_secret = secret_path + secret_name
+        response = ssm_client.get_parameters(Names=[param_name, complete_secret,
+            'inexistent_param', secret_path + 'inexistent_secret'])
+        found = response.get('Parameters')
+        notFound = response.get('InvalidParameters')
```
Comment (whummer) [inline tests/integration/test_ssm.py:85]: nit: better use `snake_case` for variable names, i.e., `not_found`

File: localstack/services/ssm/ssm_listener.py [hunk 6]
```diff
@@ -19,12 +19,44 @@ def normalize_name(param_name):
...
+    result = {'Parameters': [], 'InvalidParameters': []}
+
+    for name in names:
+        if name[0:29] == '/aws/reference/secretsmanager':
+            secret = get_secrets_information(name, name[30:])
```
Comment (whummer) [inline localstack/services/ssm/ssm_listener.py:48]: Let's avoid hardcoding the index offset (`30`) here - better:
```
name_prefix = '/aws/reference/secretsmanager'
if name.startswith(name_prefix):
    secret = get_secrets_information(name, name[len(name_prefix):])
    ...
```

**Review by whummer — COMMENTED** [review]
Thanks for this pull request @jellybeat ! Added some minor comments, can you please take a look?

As far as the Java issue is concerned - do you think you can share a small code snippet that would help reproduce this issue? Also, would be great if you could extend the tests in the Java utils repo [here](https://github.com/localstack/localstack-java-utils) to cover this functionality. Thanks

**Comment by jellybeat** [comment]
@whummer thanks for the comments, they have been addressed.
I'll provide a small java snippet to illustrate the issue when possible and see if I can extend the tests in the java repo as well once I get to know it better 👍

**Comment by whummer** [comment]
Looks great - thanks for updating the PR @jellybeat 👍

**Pull request merged**